hands-on-exercise/1-native/2-build-module.md

```diff
@@ -0,0 +1,947 @@
+
+proto-gen:
+    @echo "Generating protobuf files..."
+    @$(protoImage) sh ./scripts/protocgen.sh
+    @go mod tidy
+
+proto-format:
+    @$(protoImage) find ./ -name "*.proto" -exec clang-format -i {} \;
+
+proto-lint:
+    @$(protoImage) buf lint proto/ --error-format=json
+
+.PHONY: proto-all proto-gen proto-format proto-lint
+
+#################
+###  Linting  ###
+#################
+
+golangci_lint_cmd=golangci-lint
+golangci_version=v1.51.2
+
+lint:
+    @echo "--> Running linter"
+    @go install github.com/golangci/golangci-lint/cmd/golangci-lint@$(golangci_version)
+    @$(golangci_lint_cmd) run ./... --timeout 15m
+
+lint-fix:
+    @echo "--> Running linter and fixing issues"
+    @go install github.com/golangci/golangci-lint/cmd/golangci-lint@$(golangci_version)
+    @$(golangci_lint_cmd) run ./... --fix --timeout 15m
+
+.PHONY: lint lint-fix
+```
+
+Note how it uses the `ghcr.io/cosmos/proto-builder:0.14.0` Docker image here to run the `protocgen.sh` script. This helps with making sure all necessary software is available.
+
+### First Protobuf compilation
+
+Before you run the compilation, because you have not yet defined any Protobuf files for messages, you can comment out this `protocgen.sh` line:
+
+```diff-sh
+-   cp -r github.com/alice/checkers/* ./
++   # cp -r github.com/alice/checkers/* ./
+```
+
+Now you can run the compilation:
+
+```sh
+$ make proto-gen
+```
+
+This creates a new `api/module/v1/module.pulsar.go` file, and updates your `go.mod`. It also creates a new `go.sum`.
+
+You need a couple more Protobuf files.
+
+### Minimum Protobuf objects
+
+You are not defining your messages and queries just yet. However, you already define the facts that your module:
+
+* Has a genesis, which has a type.
+* Uses params, which also have a type.
+
+Go ahead and define them in a new `proto/alice/checkers/v1/types.proto`, taking inspiration from `minimal-module-example`:
+
+```protobuf [proto/alice/checkers/v1/types.proto]
+syntax = "proto3";
+package alice.checkers.v1;
+
+option go_package = "github.com/alice/checkers";
+
+import "cosmos_proto/cosmos.proto";
+import "gogoproto/gogo.proto";
+import "amino/amino.proto";
+
+// Params defines the parameters of the module.
+message Params { option (amino.name) = "alice/checkers/Params"; }
+
+// GenesisState is the state that must be provided at genesis.
+message GenesisState {
+  // params defines all the parameters of the module.
+  Params params = 1 [ (gogoproto.nullable) = false, (amino.dont_omitempty) = true ];
+}
+```
+
+Because you added a new Protobuf file, you need to compile it too. Uncomment the `cp` line:
+
+```diff-sh
+-   # cp -r github.com/alice/checkers/* ./
++   cp -r github.com/alice/checkers/* ./
+```
+
+And run it:
+
+```sh
+$ make proto-gen
+```
+
+<HighlightBox type="alert" title="Troubleshooting">
+
+If the process tells you that it downloads a Cosmos SDK version prior to 0.50, make sure to manually edit your `go.mod`.
```

Comment:
@xavierlepretre I did a `ctrl-c` in the moment I saw it tries to download the 0.47.5 SDK but than there was no line `github.com/cosmos/cosmos-sdk v0.47.5`. So I was a bit confused. If we include all those yaml files etc. perhaps it is better if we also include the expected version of `go.mod` before running `make proto-gen`?

Reply:
Ah yes, it is something to clear up. I am waiting for Git tags on repos.

Reply:
Done https://github.com/cosmos/developer-portal/pull/56